### PR TITLE
RSDK-8886 Setup phase

### PR DESCRIPTION
**Pull request by maximpertsov**
### Summary

* Add an optional "setup phase" to modules, that which a user-defined, module-specific setup script whenever a module is added or modified. All setup phases must execute successfully (i.e. `exit code == 0`) before viam-server caches the associated config update and proceeds with the (re)configuration.

* Setup phase should only be run once per module version.

* This feature primarily aims to support docker-ized modules that might require online connectivity during configuration.

### Non-unit testing

I uploaded a custom private module that is based on [a custom branch](https://github.com/viam-labs/wifi-sensor/tree/RSDK-8886-setup-phase) of [this module](https://github.com/viam-labs/wifi-sensor)

### TODO

* [x] FirstRun once per module version (persist between runs) - https://github.com/viamrobotics/rdk/pull/4457/commits/5b9b18485ddd3b275f7e0034be32b325193b3eac
* [x] Do some manual testing with a setup script that executes `docker` and `docker compose` commands since that is the main use-case for this feature.

### CODE REVIEW

File: module/testmodule/first_run.sh
Comment (maximpertsov) [inline module/testmodule/first_run.sh:1]: note for early reviewers: this script is currently being packaged in a module that [is defined in another repo](https://github.com/viam-labs/wifi-sensor/compare/RSDK-8886-setup-phase?expand=1). I aim to have that same setup in RDK for unit testing, hence these preemptive changes - they just aren't being utilized yet.

File: robot/impl/local_robot.go
Comment (maximpertsov) [inline robot/impl/local_robot.go:1211]: I admit it's a little ugly to `DiffConfigs` twice - once here and again [later on after we resolve deps](https://github.com/viamrobotics/rdk/pull/4457/files#diff-0b975da6011656cd3f2caf0435023573bdb5ac7c6a425e7917dd5cc9d6be63e4R1286). However, we do need to this diff earlier to get Added + Modified modules before we cache the config, so I think diffing twice is the simplest way to accomplish that.

Open to suggestions on alternatives. The one other thing I considered was delaying the cache step, but that felt like a bigger/riskier change.

File: module/modmaninterface/interface.go
Comment (maximpertsov) [inline module/modmaninterface/interface.go:30]: not ideal, but I seems I need to do this make this function available in the top-level `reconfigure` function in local_robot. 

alternatively we can type-assert the module manager in local robot to cast it to our implemented module manager (and skip setup phase for other implementations), but then unit testing becomes a little harder.

File: module/modmanager/manager.go
Comment (maximpertsov) [inline module/modmanager/manager.go:1342]: refactored to make this functionality available to `FirstRun` more conveniently.

File: module/modmaninterface/interface.go
Comment (maximpertsov) [reply module/modmaninterface/interface.go:30]: side note/idea, `Setup` feels like a better name

```suggestion
	Setup(ctx context.Context, conf config.Module) error
```

File: config/module.go
Comment (maximpertsov) [inline config/module.go:223]: Largely copypasta from `EvaluateExePath`, but allows the `first_run` script path to come from `meta.json` - we don't allow users to define the path in the module config.

File: config/module.go
Comment (maximpertsov) [inline config/module.go:271]: double-reminder for myself, this should NOT be an error - we probably do just want to move on silently if we reach this point.

File: config/module.go
Comment (maximpertsov) [inline config/module.go:269]: @abe-winter just to confirm my understanding of this code path - our code docs say the following about `NeedsSyntheticPackage`

> NeedsSyntheticPackage returns true if this is a local module pointing at a tarball

So basically this refers to module tarballs that are not coming from the registry - the user just has them available locally on the machine - is that correct?

File: module/modmanager/manager.go
Comment (maximpertsov) [inline module/modmanager/manager.go:1105]: the lint is legit and though we discussed it IRL already, I just want to point it out again in case reviewers have second thoughts about this. we could consider adding _some_ sanitation here.

File: cli/module_registry.go
Comment (cheukt) [inline cli/module_registry.go:104]: isn't this already contained in modconfig.JSONManifest?

File: robot/impl/local_robot.go
Comment (cheukt) [inline robot/impl/local_robot.go:1218]: first run should be run in parallel - let's make this a fast follow but we shouldn't lose track

File: module/modmanager/manager.go
Comment (cheukt) [inline module/modmanager/manager.go:1083]: we should be running FirstRun only once per module version, do we have that logic anywhere?

File: robot/impl/local_robot.go
Comment (maximpertsov) [reply robot/impl/local_robot.go:1218]: i had this thought also and yeah it should at least be a fast-follow if not part of this PR, since the setup phase commands we are expecting will be slow (`docker compose`, etc). the setup phase logic owned by RDK should not be mutating any shared fields so operation should be thread-safe.

i suppose we need to communicate to the user that setup phases will be run in parallel, so scripts should be thread safe. alternatively we can expose an option that dictates whether to setup concurrently or serially

File: module/modmaninterface/interface.go
Comment (cheukt) [reply module/modmaninterface/interface.go:30]: FirstRun is more consistent with everything else, wouldn't change it

File: cli/module_registry.go
Comment (maximpertsov) [reply cli/module_registry.go:104]: it is, will remove

File: robot/impl/local_robot.go
Comment (maximpertsov) [reply robot/impl/local_robot.go:1218]: discussed IRL - it's okay to have setup scripts run concurrently - we just need to document this behavior clearly. it's okay since we already do module/resource processing concurrently. the main difference, at least w.r.t resources, is that users cannot specific dependency chains with setup scripts, but they can get around this by bundling all setup logic into a single script.

File: module/modmaninterface/interface.go
Comment (maximpertsov) [reply module/modmaninterface/interface.go:30]: also note to self: modmanager test interfaces need to be updated

**Comment by cheukt** [comment]
this is also missing support for "first_run_timeout", can you add a ticket to track?

**Comment by maximpertsov** [comment]
> this is also missing support for "first_run_timeout", can you add a ticket to track?

done, added tickets for this and parallel processing and running once per module version. will address each as a fast follow

File: config/module.go
Comment (maximpertsov) [reply config/module.go:268]: changed my mind - this should be an error - but it is [gracefully handled downstream](https://github.com/viamrobotics/rdk/pull/4457/files#diff-a4308b77f384da81096eda21bf023471e80818dda86e74f9335f99a6a57d07c9R1087-R1090)

File: config/module.go
Comment (maximpertsov) [reply config/module.go]: thanks! 

@cheukt to confirm, should this be a valid path to accessing a setup script? my thinking is yes but let me know if you disagree

File: module/modmanager/manager.go
Comment (maximpertsov) [inline module/modmanager/manager.go:1089]: discussed IRL: we may want to distinguish between different kinds of errors in evaluating the first run path to determine whether to debug log or warn.

File: module/testmodule/first_run.sh
Comment (maximpertsov) [reply module/testmodule/first_run.sh:1]: decided to delete - will add testing as a follow-up

File: config/module.go
Comment (maximpertsov) [inline config/module.go:220]: instead of using the existing `.status.json` file, I opted to create a new file whose existence marks a successful setup phase run. this was driven by two factors:

* having a separate files avoids any issues of potentially corrupting the `.status.json` file
* importing `packages` in this file (to get the defined `.status.json` extension) creates a circular import

File: robot/packages/utils.go
Comment (maximpertsov) [inline robot/packages/utils.go:276]: this logic is necessary to prevent the cleanup routine from deleting the setup phase success marker file

File: robot/impl/local_robot.go
Comment (maximpertsov) [reply robot/impl/local_robot.go]: we actually don't need to diff - we can just run the setup phase for all modules in the new config

File: config/module.go
Comment (maximpertsov) [reply config/module.go:220]: > immortalize 

sounds good - will doc this now that there’s some consensus 

> extension 

it is an extension - the full file ends up being the versioned module name + this extension

File: robot/packages/utils.go
Comment (maximpertsov) [reply robot/packages/utils.go:276]: sounds good, will add docs. as for consequences - deleting the marker file results in first run potentially running again the first time a new version of the package is downloaded.

Comment (dgottlieb) [reply robot/packages/utils.go:276]: I think the consequence is more interesting? In what circumstance would we clean up this directory, but then redownload into it?

Comment (maximpertsov) [reply robot/packages/utils.go:276]: hm, i guess i don’t know all the circumstances when this happens. 

my understanding is that this directory is cleaned when a module/version is no longer in the config but otherwise everything persists. @abe-winter is this understanding correct or am i missing some subtleties.

File: module/modmanager/manager.go
Comment (cheukt) [reply module/modmanager/manager.go:1093]: if first run already succeeded, would be good to separately info log that a first_run script was detected but will not be run

File: module/modmanager/manager.go
Comment (cheukt) [inline module/modmanager/manager.go:1124]: likely a followup, but this output does not show on app.viam logs. An option would be to dump the output as part of a Infof or Errorf log

**Review by cheukt — APPROVED** [review]
tested on a pi with a custom module - great work! only runs one time

LGTM % Dan's comments

File: config/module.go
Comment (cheukt) [inline config/module.go:241]: nit: organizationally, I don't think the evaluation of first run success should be in a `EvaluateFirstRunPath` function - it feels like the function you would run before this function, since you wouldn't need to find the first run path any more at this point. However, I do see a lot of vars are shared so I can see this being fine as well

Comment (maximpertsov) [reply config/module.go:241]: > I do see a lot of vars are shared so I can see this being fine as well

yep, this is why i organized it the way you see it. your suggestion is how i would want it to be organized also. 

i will merge this as is and try to re-arrange when addressing all of the fast-follows for this feature

File: module/modmanager/manager.go
Comment (maximpertsov) [reply module/modmanager/manager.go:1124]: oh i think this is because i did the AsZap logger conversion in order to use the `With` method. that’s annoying… oh well, yeah i’ll follow up on this

File: robot/packages/utils.go
Comment (abe-winter) [reply robot/packages/utils.go:276]: > In what circumstance would we clean up this directory, but then redownload into it?

dgottlieb are you saying you know of this case? or just thinking defensively

maximpertsov speaking from memory, this cleanup logic deletes inner folders but not the directory itself (unless there's something at the bottom that deletes the whole dir when empty)

I am not now nor ever have I been particularly clear on the overall file layout of the ~/.viam dir, even though I have messed with it a lot -- I get confused btwn the data, download, and unpack dirs.

don't let me be a merge blocker, but if you want to sync + document the ~/.viam subfolders at some point, would not mind taking an hour

Comment (dgottlieb) [reply robot/packages/utils.go:276]: > dgottlieb are you saying you know of this case? or just thinking defensively

Not thinking defensively. The code written here is already being defensive. But it's not communicating the reason why.

I was just hoping (perhaps wishfully) that we could write it down. Such if this defensive code turned out to be problematic for one scenario, we'd have more confidence about the case this was meant to guard against and preserve that behavior.

I mean, I suspect* this code is referring to a scenario of:
* User configures some (module, version) to run
* The package is downloaded successfully and the module runs for some time
* The user configures away the (module, version). The package is deleted except:
  * for the meta.json file and
  * the "first run" file.

It's not clear to me what bad thing would happen if the meta.json file were to be deleted. But it does seem right to keep the "first run" file.

But that's just based on @maximpertsov 's original (pr) comment:
> this logic is necessary to prevent the cleanup routine from deleting the setup phase success marker file

It's not even obvious to me that's even what the code is doing. I don't see the code preserving certain files. I just see it deleting an entire directory.

File: config/module.go
Comment (maximpertsov) [reply config/module.go:240]: this is creating a file not a sub-directory. the marker file has the format `<module_directory>.first_run_succeeded`. 

separately, `firstRunSuccessPath` might not be a good variable name so let me change that.

File: config/module.go
Comment (maximpertsov) [reply config/module.go:247]: yeah I should be explicit about that - we make our best effort to the create the marker file - if that fails we will run the setup phase again. we can harden this as a follow-up if that's an important requirement in this system.

File: config/module.go
Comment (maximpertsov) [reply config/module.go:220]: i'm gonna rename the variable to `Suffix` instead - i guess it's a bit weird to call `.first_run_succeeded` an "extension"

**Comment by maximpertsov** [comment]
@cheukt @dgottlieb thank you for reviewing! I made a few no-op changes to address some of your comments (limited to adding documentation and renaming variables). I'm going to address everything else in a separate PR so that I don't nullify [manual testing](https://github.com/viamrobotics/rdk/pull/4457#pullrequestreview-2379028508).

Once CI passes I will merge as is.

File: robot/packages/utils.go
Comment (maximpertsov) [reply robot/packages/utils.go]: (side note - thank you for all of the context - I'm going to address this documentation in a separate PR - mainly because I don't know exactly what to document here yet)

Comment (cheukt) [reply robot/packages/utils.go]: my understanding is that the variables are not named well/haven't been renamed to suit the newer types of files we have added. Looking at a .viam dir with a tarball module right now, the module is unpacked into .viam/packages/data/module. That should correspond to a `packageTypeDir`. 

We read everything in the data/module dir with `packageDirs, err := os.ReadDir(packageTypeDirName)`. each packageDir in packageDirs can either be a directory, status.json, or .first_run_success file. In my case, there is a mod v1 directory (the extracted tarball), mod v1 status.json, and mod v1 first_run_succeeded file.

Looping through every packageDir entry, we decide whether or not to delete it based on if it is still being managed by the package manager. So if the package manager is supposed to be managing mod v2, all those previously listed files would be deleted based on the checks in L273-276. The first check checks if the file/dir name matches a module being managed, and the other two checks if the file/dir name matches a module being managed after the suffix has been removed.

The behavior now makes sense to me (if we start with v1, first_run succeeds and then we go to v2 and then back to v1, we should probably run v1's first_run again).

packageDir should probably have a different name haha

File: robot/packages/utils.go
Comment (cheukt) [inline robot/packages/utils.go:256]: I'm not sure if .status.json or .first_run_succeeded files should be at this level? .first_run_succeeded definitely not, since they are in .viam/packages/data/module, though unclear if we download other things into this top level with the corresponding .status.json file

Comment (cheukt) [reply robot/packages/utils.go:256]: based on how getSyncFileName works, I don't think .status.json files will ever be at the packages/data dir level (LocalDataDirectory joins LocalDataParentDirectory with the package name, and LocalDataParentDirectory is already `.viam/packages/data/XXX`)

**Pull request merged**
